examples/pxScene2d/src/rcvrcore/optimus.js

```diff
@@ -316,91 +321,108 @@ function Application(props) {
-            } else {
-              _this.log("failed to create window for WebApp");
-              _this.applicationClosed();
-            }
-          } else {
-            _this.log("failed to create WebApp invalid waylandObj");
-            _this.applicationClosed();
-          }
-        }, function rejection() {
-          _this.log("failed to create WebApp");
-          _this.applicationClosed();
-        });
-        this.setProperties(props);
-      }
-      else{
-        this.type = ApplicationType.NATIVE;
-        _externalApp = scene.create( {t:"wayland", parent:root, cmd:cmd, w:w, h:h, hasApi:true} );
-        this.ready = _externalApp.ready;
-        _externalApp.on("onReady", function () { _this.log("onReady"); }); // is never called
-        _externalApp.on("onClientStarted", function () { _this.log("onClientStarted"); });
-        _externalApp.on("onClientConnected", function () { _this.log("onClientConnected"); });
-        _externalApp.on("onClientDisconnected", function () { _this.log("onClientDisconnected"); }); // called on client crash
-        _externalApp.on("onClientStopped", function () { // called on client crash
-          _this.log("onClientStopped");
-          setTimeout(function () {
-            _this.destroy();
-          });
-        });
-        _externalApp.ready.then(function() {
-          _this.log("successfully created: " + _this.id);
+
+  if (!cmd) {
+    this.log('cannot create app because cmd is not set');
+    _readyReject(new Error('cmd is not set'));
+    setTimeout(function () { // app not created yet
+      _this.applicationClosed();
+    });
+  }
+  else if (!scene) {
+    this.log('cannot create app because the scene is not set');
+    _readyReject(new Error('scene is not set'));
+    setTimeout(function () { // app not created yet
+      _this.applicationClosed();
+    });
+  }
+  else if (appManager.getApplicationById(props.id)) {
+    this.log('cannot create app with duplicate ID: ' + props.id);
+    _readyReject(new Error('duplicate ID: ' + props.id));
+    setTimeout(function () { // app not created yet
+      _this.applicationClosed();
+    });
+  }
+  else if (cmd === "spark"){
+    this.type = ApplicationType.SPARK;
+    _externalApp = scene.create({t:"scene", parent:root, url:uri, serviceContext:serviceContext});
+    _externalApp.on("onReady", function () { _this.log("onReady"); }); // is never called
+    _externalApp.on("onClientStarted", function () { _this.log("onClientStarted"); }); // is never called
+    _externalApp.on("onClientConnected", function () { _this.log("onClientConnected"); }); // is never called
+    _externalApp.on("onClientDisconnected", function () { _this.log("onClientDisconnected"); }); // is never called
+    _externalApp.on("onClientStopped", function () { _this.log( "onClientStopped"); }); // is never called
+    _externalApp.ready.then(function() {
+      _this.log("successfully created Spark app: " + _this.id);
+      _readyResolve();
```

Comment:
Let's check to ensure _readyResolve is defined before calling it (also in other usages).

Reply:
It is always defined bcs promise executor is called immediately.